File: python/2022/25_full_of_hot_air/aoc202225.py
"""AoC 25, 2022: Full of Hot Air."""

# Standard library imports
import pathlib
import sys

TO_SNAFU = {-2: "=", -1: "-", 0: "0", 1: "1", 2: "2"}
FROM_SNAFU = {snafu: dec for dec, snafu in TO_SNAFU.items()}


def parse_data(puzzle_input):
    """Parse input."""
    return puzzle_input.split("\n")


def part1(fuel):
    """Solve part 1."""
    return to_snafu(sum(from_snafu(number) for number in fuel))


def part2(data):
    """There is no part 2."""


def from_snafu(snafu):
    """Convert a number from SNAFU to decimal.

    ## Examples:

    >>> from_snafu("22")
    12
    >>> from_snafu("2022")
    262
    >>> from_snafu("1=11-2")
    2022
    """
    return sum(
        5**power * FROM_SNAFU[digit] for power, digit in enumerate(snafu[::-1])
    )


def to_snafu(decimal):
    """Convert a number from decimal to SNAFU.

    ## Examples:

    >>> to_snafu(13)
    '1=='
    >>> to_snafu(1234)
    '20-2-'
    """
    snafu = []
    while decimal > 0:
        decimal, digit = divmod(decimal + 2, 5)
        snafu.append(TO_SNAFU[digit - 2])
    return "".join(snafu[::-1])


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2022/25_full_of_hot_air/aoc202225.ex
defmodule AOC2022.Day25 do
  @moduledoc """
  Advent of Code 2022, day 25: Full of Hot Air.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input), do: puzzle_input |> String.split("\n", trim: true)

  @doc """
  Solve part 1.
  """
  def part1(fuel), do: fuel |> Enum.map(&from_snafu/1) |> Enum.sum() |> then(&to_snafu/1)

  @doc """
  There is no part 2.
  """
  def part2(_), do: nil

  @doc """
  Convert from SANFU to decimal.

  ## Examples:

      iex> from_snafu("1=")
      3
      iex> from_snafu("111")
      31
      iex> from_snafu("1121-1110-1=0")
      314159265
  """
  def from_snafu(snafu) do
    snafu
    |> String.to_charlist()
    |> Enum.map(&digit_from_snafu/1)
    |> Enum.reverse()
    |> Enum.with_index()
    |> Enum.map(fn {snafu, power} -> snafu * 5 ** power end)
    |> Enum.sum()
  end

  @doc """
  Convert one SNAFU digit to decimal.

  ## Examples:

      iex> digit_from_snafu(?2)
      2
      iex> digit_from_snafu(?-)
      -1
  """
  def digit_from_snafu(?2), do: 2
  def digit_from_snafu(?1), do: 1
  def digit_from_snafu(?0), do: 0
  def digit_from_snafu(?-), do: -1
  def digit_from_snafu(?=), do: -2

  @doc """
  Convert from decimal to SNAFU.

  ## Examples:

      iex> to_snafu(12)
      "22"
      iex> to_snafu(13)
      "1=="
      iex> to_snafu(123)
      "100="
      iex> to_snafu(12345)
      "1-0---0"
  """
  def to_snafu(number), do: to_snafu(number, [])
  def to_snafu(0, snafu), do: List.to_string(snafu)

  def to_snafu(number, suffix),
    do: to_snafu(div(number + 2, 5), [digit_to_snafu(rem(number + 2, 5) - 2) | suffix])

  @doc """
  Convert one digit to SNAFU.

  ## Examples:

      iex> digit_to_snafu(1)
      ?1
      iex> digit_to_snafu(-2)
      ?=
  """
  def digit_to_snafu(2), do: ?2
  def digit_to_snafu(1), do: ?1
  def digit_to_snafu(0), do: ?0
  def digit_to_snafu(-1), do: ?-
  def digit_to_snafu(-2), do: ?=

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
